dashboard/app/controllers/scripts_controller.rb

```diff
@@ -483,4 +483,14 @@ def get_rollup_resources
     canonical_path = Services::Courses.canonical_path(request.fullpath, unit_name_or_id)
     redirect_to canonical_path unless canonical_path == request.fullpath
   end
+
+  # Authorize in a separate before_action rather than using CanCan's authorize_resource
+  # After the URL restructuring from /s/... to /courses/.../units/... the selected unit
+  # was no longer being authorized, and instead the Unit model was being authorized, leading
+  # some users to have access to certain courses they shouldn't have access to (see
+  # TEACH-1975 for more details).This solves the issue by explicitly calling authorize!
+  # on the @script if it is set.
+  private def authorize_script
+    authorize! params[:action].to_sym, @script || Unit
```

Comment:
I think I just ran into a similar issue on my UI test PR, which involves some similar logic in the script levels controller. I don't see how this can work in all cases because cancan doesn't know about the published state of the UnitGroup, which is (should be?) a factor in whether the unit is visible to a user via `show`.